hypershift-operator/controllers/nodepool/capi.go

```diff
@@ -0,0 +1,1339 @@
+	err := c.Get(ctx, client.ObjectKeyFromObject(md), md)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("error getting MachineDeployment: %w", err)
+	}
+	if md.DeletionTimestamp != nil {
+		return nil
+	}
+	err = c.Delete(ctx, md)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("error deleting MachineDeployment: %w", err)
+	}
+	return nil
+}
+
+func (c *CAPI) pauseMachineDeployment(ctx context.Context) error {
+	md := c.machineDeployment()
+	err := c.Get(ctx, client.ObjectKeyFromObject(md), md)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("error getting MachineDeployment: %w", err)
+	}
+	if md.Annotations == nil {
+		md.Annotations = make(map[string]string)
+	}
+	// FIXME: In future we may want to use the spec field instead
+	// https://github.com/kubernetes-sigs/cluster-api/issues/6966
+	md.Annotations[capiv1.PausedAnnotation] = "true"
+	return c.Update(ctx, md)
+}
+
+func deleteMachineSet(ctx context.Context, c client.Client, ms *capiv1.MachineSet) error {
+	err := c.Get(ctx, client.ObjectKeyFromObject(ms), ms)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("error getting MachineSet: %w", err)
+	}
+	if ms.DeletionTimestamp != nil {
+		return nil
+	}
+	err = c.Delete(ctx, ms)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("error deleting MachineSet: %w", err)
+	}
+	return nil
+}
+
+func (c *CAPI) Pause(ctx context.Context) error {
+	// Pause MachineSet
+	if err := c.pauseMachineSet(ctx); err != nil {
+		return fmt.Errorf("error pausing MachineSet: %w", err)
+	}
+
+	// Pause MachineDeployment
+	if err := c.pauseMachineDeployment(ctx); err != nil {
+		return fmt.Errorf("error pausing MachineDeployment: %w", err)
+	}
+
+	return nil
+}
+
+func (c *CAPI) pauseMachineSet(ctx context.Context) error {
+	ms := c.machineSet()
+	err := c.Get(ctx, client.ObjectKeyFromObject(ms), ms)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("error getting MachineSet: %w", err)
+	}
+	if ms.Annotations == nil {
+		ms.Annotations = make(map[string]string)
+	}
+	// FIXME: In future we may want to use the spec field instead
+	// https://github.com/kubernetes-sigs/cluster-api/issues/6966
+	// TODO: Also for paused to be complete we will need to pause all MHC if autorepair
+	// is enabled and remove the autoscaling labels from the MachineDeployment / Machineset
+	ms.Annotations[capiv1.PausedAnnotation] = "true"
+	return c.Update(ctx, ms)
+}
+
+func deleteMachineHealthCheck(ctx context.Context, c client.Client, mhc *capiv1.MachineHealthCheck) error {
+	err := c.Get(ctx, client.ObjectKeyFromObject(mhc), mhc)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("error getting MachineHealthCheck: %w", err)
+	}
+	if mhc.DeletionTimestamp != nil {
```

Comment:
why do we need to fetch the object and check the DeletionTimestamp first? isn't `Delete` a no-op if the object is already deleting?

Reply:
This is just the business logic of the original function. I'd prefer to keep the scope of this PR small to reduce risk of regression and let any follow up iteration be test driven now we have decent coverage